nemo/collections/asr/data/audio_to_text.py

```diff
@@ -208,6 +226,125 @@ def expand_audio_filepaths(audio_tar_filepaths, shard_strategy: str, world_size:
+
+    num_store_manifests = sum([is_store_path(f) for f in manifest_filepaths])
+
+    if num_store_manifests > 0:
+        # Local utility function
+        def cache_data(manifest_filepaths, cache_audio, num_workers, max_num_workers):
+            """Cache manifests and audio data from object store.
+            """
+            if num_workers is None:
+                num_workers = os.cpu_count() - 1
+            num_workers = min(num_workers, max_num_workers)
+
+            for manifest_file in manifest_filepaths:
+                if is_store_path(manifest_file):
+                    logging.info('Cache manifest file: %s', manifest_file)
+                    cached_manifest_file = StoreObject(manifest_file).get()
+                    logging.info('Cached at: %s', str(cached_manifest_file))
+
+                    if cache_audio:
+                        logging.info('Cache audio from manifest file: %s', manifest_file)
+                        # Assumes that manifest is using relative paths
+                        manifest_dir = os.path.dirname(manifest_file)
+                        # Prepare all store objects
+                        audio_objects = []
+                        with open(cached_manifest_file, 'r') as f:
+                            for line in f:
+                                item = json.loads(line)
+                                store_path = os.path.join(manifest_dir, item['audio_filepath'])
+                                audio_objects.append(StoreObject(store_path=store_path))
+
+                        if num_workers is not None and num_workers > 1:
+                            logging.debug('Using multiprocessing with num_workers: %d.', num_workers)
+                            with multiprocessing.Pool(processes=num_workers) as p:
+                                result = list(tqdm(p.imap(store_object_get, audio_objects), total=len(audio_objects)))
+                        else:
+                            logging.debug('Using a single process.')
+                            result = []
+                            for audio_object in tqdm(audio_objects):
+                                result.append(audio_object.get() is not None)
+
+                        if any(result) == False:
+                            raise RuntimeError('Some files not downloaded successfully')
+                        logging.info('Caching complete')
+
+                else:
+                    # Nothing to do here
+                    logging.debug('Manifest is not on a store: %s', manifest_file)
+
+        # Handle distributed environment
+        if shared_cache is None:
+            shared_cache = is_store_cache_shared()
+
+        if shared_cache:
+            logging.debug('Cache is shared among nodes.')
+            is_rank_zero = is_global_rank_zero()
+        else:
+            logging.debug('Cache is not shared among nodes.')
+            local_rank = int(os.environ.get("LOCAL_RANK", 0))
```

Comment:
If multi gpu is setup, but local rank is not set correctly, all ranks will be set to 0. Do we want to catch this behaviour and raise warning?

Reply:
Oh its done below, but all of them will be rank 0. I guess its fine.